src/main/java/edu/uwaterloo/ece658/service/DataHandlerService.java

```diff
@@ -0,0 +1,160 @@
+        photo.setS3Key(S3key);
+        photo.setMd5(Md5);
+        photo.setSize(photoSize);
+        photo.setUploadedTime(uploadDate);
+        photo.setIsPrivate(isPrivatePhoto);
+        photoFacade.create(photo);
+        return photo;
+    }
+    
+    private Tag createNewTag(String tagName) {
+        Tag tag = new Tag();
+        tag.setName(tagName);
+        tagFacade.create(tag);
+        return tag;
+    }
+    
+    private Tag initializeTag(String tagName) {
+        Tag tag = tagFacade.getNormalTagByName(tagName);
+        if (tag == null) {
+            tag = createNewTag(tagName);
+        }
+        return tag;
+    }
+    
+    private List<Tag> initializeTags(List<String> tagNames) {
+        List<Tag> tags = new ArrayList<>();
+        
+        for (String tagName : tagNames) {
+            tags.add(initializeTag(tagName));
+        }
+        return tags;
+    }
+    
+    public boolean updateDatabaseWithPhotoInformation(
+            String username, 
+            String S3key,
+            String Md5,
+            Integer photoSize, 
+            Date uploadDate,
+            boolean isPrivatePhoto,
+            List<String> tagNames)
+    {
+        // store key in table with tags
+        List<Tag> tags = initializeTags(tagNames);
+        User user = userFacade.findUserByUserName(username);
+        Photo photo = initializePhoto(
+                        S3key,
+                        Md5,
+                        photoSize,
+                        uploadDate,
+                        isPrivatePhoto);
+        
+        photo.setTags(tags);
+        photo.addUser(user);
+        
+        user.addUploadedPhoto(photo);
+        userFacade.edit(user);
+        
+        for (Tag tag : tags) {
+            tag.addPhoto(photo);
+            tagFacade.edit(tag);
+        }
+        
+        // call notify to send messages to users
+        return subscriptionService.notifyUsers(tags, photo);
+    }
+    
+    public void removeTagFromPhoto(String username, String S3key, String tagName) {
+        Photo photo = photoFacade.getPhotoByS3Key(S3key);
+        User user = userFacade.findUserByUserName(username);
+        Tag tag = tagFacade.getNormalTagByName(tagName);
+        
+        if (photo.getUploadedUsers().contains(user)) {
+            photo.getTags().remove(tag);
+            tag.getPhotosUnderThisTag().remove(photo);
+            tagFacade.edit(tag);
+            photoFacade.edit(photo);
+        }
+    }
+    
+    public void addTagToPhoto(String username, String S3key, String tagName) {
+        Photo photo = photoFacade.getPhotoByS3Key(S3key);
+        User user = userFacade.findUserByUserName(username);
+        Tag tag = initializeTag(tagName);
+        
+        if (photo.getUploadedUsers().contains(user)) {
+            photo.addTag(tag);
+            tag.addPhoto(photo);
+            tagFacade.edit(tag);
+            photoFacade.edit(photo);
+        }
+    }
+    

```

Comment:
Adding a method called `updatePhotoTags(String userName, String s3Key, List<String> newTags)` would be good so that front end only needs to call this method instead of iterating over all the tags of a photo and calling `removeTagToPhoto` method and `addTagToPhoto` in the controller.  `updatePhotoTags(String s3Key, List<String> tags)`  internally handles correctly the addition and deletion of `Tag`: if the `newTags` contains existing tags, for those tags do nothing; But if old tags is not included in `newTags` call `removeTagToPhoto` ; Otherwise, call `addTagToPhoto`.